src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,914 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with OpenSlide. If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+ * DICOM (.dcm) support
+ */
+
+#include "openslide-private.h"
+#include "openslide-decode-jpeg.h"
+#include "openslide-hash.h"
+
+#include <glib.h>
+
+#include <dicom/dicom.h>
+
+enum image_format {
+  FORMAT_UNKNOWN,
+  FORMAT_JPEG,
+};
+
+struct dicom_ops_data {
+  char *dirname;
+  int64_t tile_size;
+};
+
+struct dicom_file {
+  char *filename;
+
+  GMutex lock;
+  DcmFilehandle *filehandle;
+  DcmDataSet *metadata;
+  DcmDataSet *meta;
+  DcmBOT *bot;
+};
+
+struct dicom_level {
+  struct _openslide_level base;
+  struct _openslide_grid *grid;
+
+  enum image_format image_format;
+  int64_t image_width;
+  int64_t image_height;
+  int64_t tile_w;
+  int64_t tile_h;
+  guint32 num_frames;
+  int64_t tiles_across;
+  int64_t tiles_down;
+
+  struct dicom_file *file;
+};
+
+struct associated {
+  struct _openslide_associated_image base;
+
+  const char *name;
+  struct dicom_file *file;
+};
+
+// a set of allowed image types for a class of image
+struct allowed_types {
+  const char ***types;
+  int n_types;
+};
+
+// the ImageTypes we allow for pyr levels
+static const char *original_types[4] = {
+  "ORIGINAL", "PRIMARY", "VOLUME", "NONE"
+};
+static const char *resampled_types[4] = {
+  "DERIVED", "PRIMARY", "VOLUME", "RESAMPLED"
+};
+static const char **level_type_strings[] = {
+  original_types,
+  resampled_types
+};
+
+static const struct allowed_types level_types = {
+  level_type_strings,
+  sizeof(level_type_strings) / sizeof(level_type_strings[0])
+};
+
+// the ImageTypes we allow for associated images
+static const char *label_types[4] = {
+  "ORIGINAL", "PRIMARY", "LABEL", "NONE",
+};
+static const char *overview_types[4] = {
+  "ORIGINAL", "PRIMARY", "OVERVIEW", "NONE",
+};
+static const char **associated_type_strings[] = {
+  label_types,
+  overview_types
+};
+static const struct allowed_types associated_types = {
+  associated_type_strings,
+  sizeof(associated_type_strings) / sizeof(associated_type_strings[0])
+};
+
+/* The DICOM UIDs and fields we check.
+ */
+static const char VLWholeSlideMicroscopyImageStorage[] =
+  "1.2.840.10008.5.1.4.1.1.77.1.6";
+static const char SeriesInstanceUID[] = "SeriesInstanceUID";
+static const char TotalPixelMatrixColumns[] = "TotalPixelMatrixColumns";
+static const char TotalPixelMatrixRows[] = "TotalPixelMatrixRows";
+static const char Columns[] = "Columns";
+static const char Rows[] = "Rows";
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(DcmFilehandle, dcm_filehandle_destroy)
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(DcmDataSet, dcm_dataset_destroy)
+
+static void set_gerror_from_dcm_error(GError **err, DcmError *dcm_error) {
```

Comment:
For consistency with elsewhere in the library, let's use `propagate` as the verb in these function names.

Reply:
OK, I swapped the names.